source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2534 @@
+  Test files SHOULD NOT expect events from multiple specs (e.g. command
+  monitoring *and* SDAM events) for a single client. See
+  `Mixing event types in observeEvents and expectEvents`_ for more
+  information.
+
+.. _test_outcome:
+
+- ``outcome``: Optional array of documents. Each document will specify expected
+  contents of a collection after all operations have been executed. The list of
+  documents therein SHOULD be sorted ascendingly by the ``_id`` field to allow
+  for deterministic comparisons.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `collectionData`_.
+
+
+operation
+~~~~~~~~~
+
+An operation to be executed as part of the test.
+
+The structure of this object is as follows:
+
+.. _operation_name:
+
+- ``name``: Required string. Name of the operation (e.g. method) to perform on
+  the object.
+
+.. _operation_object:
+
+- ``object``: Required string. Name of the object on which to perform the
+  operation. This should correspond to either an `entity`_ name (for
+  `Entity Test Operations`_) or "testRunner" (for `Special Test Operations`_).
+  If the object is an entity, The YAML file SHOULD use an `alias node`_ for its
+  ``id`` field (e.g. ``object: *collection0``).
+
+.. _operation_arguments:
+
+- ``arguments``: Optional object. Map of parameter names and values for the
+  operation. The structure of this object will vary based on the operation.
+  See `Entity Test Operations`_ and `Special Test Operations`_.
+
+  The ``session`` parameter is handled specially (see `commonOptions_session`_).
+
+.. _operation_expectError:
+
+- ``expectError``: Optional `expectedError`_ object. One or more assertions for
+  an error expected to be raised by the operation.
+
+  This field is mutually exclusive with
+  `expectResult <operation_expectResult_>`_ and
+  `saveResultAsEntity <operation_saveResultAsEntity_>`_.
+
+.. _operation_expectResult:
+
+- ``expectResult``: Optional mixed type. A value corresponding to the expected
+  result of the operation. This field may be a scalar value, a single document,
+  or an array of documents in the case of a multi-document read.  Test runners
+  MUST follow the rules in `Evaluating Matches`_ when processing this assertion.
+
+  This field is mutually exclusive with `expectError <operation_expectError_>`_.
+
+.. _operation_saveResultAsEntity:
+
+- ``saveResultAsEntity``: Optional string. If specified, the actual result
+  returned by the operation (if any) will be saved with this name in the
+  `Entity Map`_.  The test runner MUST raise an error if the name is already in
+  use. If the operation does not return a value (e.g. void method), the test
+  runner MUST store an empty value (e.g. ``null``) for the entity such that the
+  name will still be defined in the entity map.
+
+  This field is mutually exclusive with `expectError <operation_expectError_>`_.
+
+  This is primarily used for creating a `changeStream`_ entity from the result
```

Comment:
Quick question here: is there a use case in mind for storing any other kind of result besides a change stream? More specifically, would it ever be useful/expected to save the result from a regular entity operation (e.g. a CRUD operation on a `Collection` entity)? Context: I'm representing the various types of entities possible with an enum, and we're not sure if it's necessary to include a catch-all variant for results from operations that aren't one of the defined entity types above.

Reply:
There is nothing in the test format prohibiting saving an arbitrary result in the entity map. I can imagine this might be more useful down the line with saving the result of a `findOne` operation; however, we'd likely also need to introduce some special test operations to work with such a value (e.g. attempt to match an arbitrary entity map value with an expected document).

I think you'd be OK with restricting things now and throwing an exception if you attempted to store an unsupported type in the map. That would be defensive and alert you to some test attempting to do so down the line.

Reply:
Hey, sorry to jump in here, but I'm a bit worried by the idea that arbitrary types can be stored in the entity map. This creates a significant burden on languages like Rust, where we can't easily define a map with arbitrary values in it. While we can do so in a fairly straightforward manner for a given set of types, adding new types to this set requires making making code changes everywhere that the map is used. We've dealt with similar issues like this in the past with our BSON library; because users are pattern matching to check which type a given BSON value is, adding a new type that they have to check is a breaking change, and we treat it accordingly in our versioning. I feel pretty strongly that adding a new type of entity which did not exist before should require a major version bump in the format, as it will constitute a significant breaking change to the Rust driver's implementation of the spec test format regardless of how we implement it.

Reply:
> I feel pretty strongly that adding a new type of entity which did not exist before should require a major version bump in the format

I think a major version bump is a bit strong in this case, and would rather see that reserved for things like renaming/removing a field or significant changes to the test runner implementation (e.g. how observed events are collated). If the test runner is designed defensively and raises an error when encountering an unsupported type in `saveResultAsEntity`, I would argue that supporting a new type in the entity map is more akin to introducing a new field in the test format that a current runner might not know how to interpret and might ignore (unless they go out of their way to raise errors on unknown fields). New fields in the test format would be a minor version bump.

If supporting a new minor version of the test format requires a major version bump in a language project (assuming the "unified test runner" was its own, versioned Rust package), I don't see the problem with that.

As a compromise, we could define supported entity types explicitly and provide guidance for test runners to defensively raise an error if they don't support a type for whatever reason. In that case, I think the explicit list would be limited to the current types in `createEntities` and additionally a change stream.

Reply:
I think explicitly listing the currently supported entity types would be a useful addition to the spec. I also think part of the difficulty for Rust is that from what I understand, arbitrary entities can appear virtually anywhere in a test file without any regard for what type the driver might expect in that location. This essentially forces us into dynamic typing, where we have to manually check each value to see if it's a type we can use or not. This requires a significant amount of boilerplate, and the resulting noise makes the code much harder to read and understand. One thing that might ease the burden a bit is if usage of entity map lookups also required a tag indicating the type of the value they were looking up, which would allow us to use a separate map for each type of entity.

Reply:
> One thing that might ease the burden a bit is if usage of entity map lookups also required a tag indicating the type of the value they were looking up, which would allow us to use a separate map for each type of entity.

Is it not possible for the rust runner to create a lookup table for this purpose when entities are being created? At the time of creation, you will know the name that will be used to perform the lookup and also the type of the entity being created so constructing this lookup table should be straightforward.

Reply:
If we look up the type first, then we still have to write code that deals with all of the cases where the type is not what we expect; that doesn't reduce the amount of boilerplate needed in every location we look up an entity.

> spare other driver languages that don't face this limitation from the extra bookkeeping cost of a type field

I don't think the "cost" of ignoring a field that doesn't affect how a driver's harness behaves is very comparable to having to manually enumerate all possible types an entity can be every single time one is looked up and then handle each case.

Reply:
@saghm: First question to get back to the issue at hand: if we explicitly list the supported types for the entity map and document that supporting new entity types warrants a minor version bump for the test format, does that allow us to resolve this issue?

----

On a separate note pertaining to how this might be implemented, I had the following in mind for C:

* map of structs indexed by strings, where each struct shares a base type (e.g. "entityPointer") and stores a type discriminator/code and a typed pointer to the corresponding entity.
* list of structs, where each also has a string containing the entity ID (if libmongoc doesn't have a hash/map structure available to use)

Either way, there is a finite number of types to support and a defensive programming approach would just have the test runner raise an error if if encounters a type it doesn't know how to handle. Feel free to point out if/why that doesn't apply to Rust.

A cursory Google search turned up [Inserting values of multiple types in Rust's HashMap](https://www.simonewebdesign.it/rust-hashmap-insert-values-multiple-types/), wherein it seems like the second approach corresponds to what I'm suggesting above and the third (multiple, typed maps) might be closer to what @prashantmital was suggesting.

Reply:
> if we explicitly list the supported types for the entity map and document that supporting new entity types warrants a minor version bump for the test format, does that allow us to resolve this issue?

With only a minor version bump, I'd still have reservations about the current design.

> On a separate note pertaining to how this might be implemented, I had the following in mind for C

I don't think this addresses my concerns at all. The issue isn't building the map, it's looking up values which can be any type in virtually any location in the format. We can't just forcibly cast the value to the type we expect; the only way the Rust compiler will let us obtain a value of the correct type is by branching and considering all possible types it could be and then acting accordingly based on that. It requires a lot of boilerplate which a slight change in the format is equipped to get rid of at no cost to other language implementations. While it would make the test format a bit more verbose, I think the goal should be to optimize for ease of implementation, not terseness of the format.

> A cursory Google search turned up Inserting values of multiple types in Rust's HashMap, wherein it seems like the second approach corresponds to what I'm suggesting above and the third (multiple, typed maps) might be closer to what @prashantmital was suggesting

The issue with the second approach is that it assumes all usages of the values will be identical. In the example they give, all they're doing with the values is printing them out, and it's easy to define an interface that contains sufficient behavior to print out any value that implements it. This isn't nearly as straightforward for the values laid out in this spec; there's no common set of behavior between a collection and a changestream that we can abstract over. The third approach has the issues I laid out in my previous comment, which are similar to the ones I described if we used the C approach.

Reply:
Based on discussion offline, it isn't feasible to avoid breaking changes to the test runner in minor version updates to the spec, so I'll resolve this now